In [1]:
# Import the required variables
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
def warn(*args, **kwargs): pass
import warnings
warnings.warn = warn

In [3]:
def num_plot(df, col, title, symb):
    fig, ax = plt.subplots(2, 1, sharex=True, figsize=(8,5),gridspec_kw={"height_ratios": (.2, .8)})
    ax[0].set_title(title,fontsize=18)
    sns.boxplot(x=col, data=df, ax=ax[0])
    ax[0].set(yticks=[])
    sns.histplot(x=col, data=df, ax=ax[1])
    ax[1].set_xlabel(col, fontsize=16)
    plt.axvline(df[col].mean(), color='darkgreen', linewidth=2.2, label='mean=' + str(np.round(df[col].mean(),1)) + symb)
    plt.axvline(df[col].median(), color='red', linewidth=2.2, label='median='+ str(np.round(df[col].median(),1)) + symb)
    plt.axvline(df[col].mode()[0], color='purple', linewidth=2.2, label='mode='+ str(df[col].mode()[0]) + symb)
    plt.legend(bbox_to_anchor=(1, 1.03), ncol=1, fontsize=17, fancybox=True, shadow=True, frameon=True)
    plt.tight_layout()
    plt.show()

In [4]:
sns.set_style("whitegrid")
sns.despine()
plt.style.use("seaborn-whitegrid")
plt.rc("figure", autolayout=True)
plt.rc("axes", labelweight="bold", labelsize="large", titleweight="bold", titlesize=14, titlepad=10)

<Figure size 432x288 with 0 Axes>

In [8]:
# Import the required variables
import pandas as pd
import numpy as np

# Read the data
Res_Nonhydro_Capita = pd.read_excel ("../data/processed/Res_Nonhydro_Capita.xlsx", index_col = 0)
SDGs_Ranks_2021 = pd.read_excel ("../data/processed/SDGs_Ranks_2021.xlsx", index_col = 0)
SDGs_Scores_2021 = pd.read_excel ("../data/processed/SDGs_Scores_2021.xlsx", index_col = 0)

# Only consider the 2021 year
Res_Nonhydro_Capita_2021 = Res_Nonhydro_Capita[[2021]]
Res_Nonhydro_Capita_2021= Res_Nonhydro_Capita_2021.rename(columns={ 2021: 'Res_Nonhydro_Capita_2021'})

# Do not consider the first column
SDGs_Ranks_2021 = SDGs_Ranks_2021.iloc[:,1:]
SDGs_Scores_2021 = SDGs_Scores_2021.iloc[:,1:]

# Append the target to the Datasets
SDGs_Ranks_2021 = pd.concat([SDGs_Ranks_2021, Res_Nonhydro_Capita_2021], axis=1)
SDGs_Scores_2021 = pd.concat([SDGs_Scores_2021, Res_Nonhydro_Capita_2021], axis=1)

# Dicotomizing Res_Nonhydro_Capita_2021 in Wealthy and Unwealthy Countries
median = Res_Nonhydro_Capita_2021['Res_Nonhydro_Capita_2021'].median()
Res_Nonhydro_Capita_2021['Res_Nonhydro_Capita_2021'] = Res_Nonhydro_Capita_2021.Res_Nonhydro_Capita_2021.map(lambda x: 1 if x >= median else 0)
Res_Nonhydro_Capita_2021 = Res_Nonhydro_Capita_2021.rename(columns={'Res_Nonhydro_Capita_2021': 'Wealthy'})

# Append the target column "Wealthy"
SDGs_Scores_2021 = pd.concat([SDGs_Scores_2021, Res_Nonhydro_Capita_2021], axis=1)

In [12]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

X = SDGs_Scores_2021.drop('Wealthy',1)
y = SDGs_Scores_2021['Wealthy']

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [16]:
from xgboost import XGBRegressor
from sklearn.metrics import accuracy_score

ss = StandardScaler()
X_train_scaled = ss.fit_transform(X_train)
X_test_scaled = ss.transform(X_test)

maxComponents = [ 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17]
list_n_neighbors = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17]

n = 0
for componentsi in maxComponents:
    pca = PCA(n_components=componentsi, random_state=42)
    pca.fit(X_train_scaled)
    X_train_scaled_pca = pca.transform(X_train_scaled)
    X_test_scaled_pca = pca.transform(X_test_scaled)
    for n_neighborsi in list_n_neighbors:
        kn_clas = KNeighborsClassifier(n_neighbors = n_neighborsi)
        kn_clas.fit(X_train_scaled_pca, y_train)
        predictions = kn_clas.predict(X_test_scaled_pca)
        if (n == 0): # The first time initialize
            Accuracy_Best = accuracy_score(y_test, predictions)
            n_neighbors_Best = n_neighborsi
            max_components_Best = componentsi
        else:
            Accuracy_New = accuracy_score(y_test, predictions)
            if ( Accuracy_Best < Accuracy_New):
                Accuracy_Best = Accuracy_New
                n_neighbors_Best = n_neighborsi
                max_components_Best = componentsi
            else:
                pass
        n +=1
print('Accuracy_Best', Accuracy_Best, 
        'max_components_Best', max_components_Best,
        'n_estimators', n_neighbors_Best)

Accuracy_Best 0.8214285714285714 max_components_Best 1 n_estimators 3


In [18]:
pipeline = Pipeline(steps = [
    ("scaler", StandardScaler()), # primero escalo
    ("pca", PCA()), # segundo aplica PCA 
    ("kneighborsclassifier", KNeighborsClassifier()) # Despues un KNeighborsClassifier
])

pipeline_param = {
    'pca__n_components' :  [1],
    'pca__random_state' :  [42],
    'kneighborsclassifier__n_neighbors' :  [3]
}

gs_pipeline = GridSearchCV(pipeline,
                            pipeline_param,
                            cv = 10,
                            scoring = 'accuracy',
                            verbose = 1, # mensajes del output
                            n_jobs = -1)

search = GridSearchCV(pipeline, pipeline_param, cv=5).fit(X_train, y_train)

print("Train: Coeficiente de determinacion de la predicción:", search.best_estimator_.score(X_train, y_train))
print("Test: Coeficiente de determinacion de la predicción:", search.best_estimator_.score(X_test, y_test))

Train: Coeficiente de determinacion de la predicción: 0.8807339449541285
Test: Coeficiente de determinacion de la predicción: 0.8214285714285714


# ABAJO ANTOGUO

In [35]:
pipeline = Pipeline(steps = [
    ("scaler", StandardScaler()), # primero escalo
    ("pca", PCA()), # segundo aplica PCA 
    ("kneighborsclassifier", KNeighborsClassifier()) # Despues un KNeighborsClassifier
])

pipeline_param = {
    'pca__n_components' :  list(range(1,17)),
    'kneighborsclassifier__n_neighbors' :  list(range(1,50))
}

gs_pipeline = GridSearchCV(pipeline,
                            pipeline_param,
                             cv = 10,
                            scoring = 'accuracy',
                            verbose = 1, # mensajes del output
                            n_jobs = -1)

grids = {"gs_pipeline": gs_pipeline}

for nombre, grid_search in grids.items():
    grid_search.fit(X_train, y_train)

Fitting 10 folds for each of 784 candidates, totalling 7840 fits


In [36]:
print(gs_pipeline.best_score_)
print(gs_pipeline.best_params_)
print(gs_pipeline.best_estimator_)

0.889090909090909
{'kneighborsclassifier__n_neighbors': 11, 'pca__n_components': 2}
Pipeline(steps=[('scaler', StandardScaler()), ('pca', PCA(n_components=2)),
                ('kneighborsclassifier', KNeighborsClassifier(n_neighbors=11))])


In [37]:
print("Accuracy train", gs_pipeline.best_estimator_.score(X_train, y_train))
print("Accuracy test", gs_pipeline.best_estimator_.score(X_test, y_test))

Accuracy train 0.8807339449541285
Accuracy test 0.8214285714285714


## MEJOR MODEL
* n_neighbors 11
* n_components 2
* Accuracy train 0.8807339449541285
* Accuracy test 0.8214285714285714

In [187]:
import pickle

with open('model_v1', 'wb') as archivo_salida:
    pickle.dump(gs_pipeline.best_estimator_, archivo_salida)

In [188]:
with open('model_v1', 'rb') as archivo_entrada:
    loaded_model = pickle.load(archivo_entrada)

In [189]:
loaded_model.score(X_test,y_test)

0.8214285714285714

In [171]:
test_csv = pd.concat([X_test, y_test], axis=1)
test_csv

SDG1       SDG2       SDG3       SDG4       SDG5  \
Country                                                                         
Portugal                  99.3815  64.104500  90.954769  99.003333  83.272750   
Poland                    99.4655  68.159875  85.269923  98.298667  72.180000   
Belize                    55.3460  59.362571  72.339308  77.694333  56.378250   
Central African Republic   0.0000  43.645500  16.026071   9.207000  30.888250   
Turkey                    99.6240  65.068000  82.928000  96.273000  46.329750   
Italy                     98.0600  71.018625  94.043000  96.708333  72.829000   
Ireland                   99.7950  75.894875  93.231846  99.738708  76.390250   
Sri Lanka                 91.5440  62.129625  77.814571  97.413667  52.004250   
Vietnam                   95.1425  71.032625  72.467429  97.107667  72.203250   
Mauritania                70.7420  43.194571  41.840154  39.399000  32.273750   
Malta                     99.8285  67.044125  91.481308  99.218333  62.245500   
Finland                   99.8480  64.164625  94.769615  98.160708  91.371000   
Jordan                    95.3335  59.558375  75.677308  71.631333  42.966750   
Jamaica                   83.3785  55.846250  79.974071  82.612000  78.205000   
Comoros                   49.6740  45.817333  53.262071  51.885667  38.553500   
El Salvador               90.1590  54.954875  78.271071  79.705000  68.514500   
Chad                      20.7300  32.216286  21.536500  14.464667  26.821500   
Bulgaria                  97.5070  66.347250  79.553000  67.553000  72.472000   
Venezuela, RB             13.2440  43.225857  65.286786  82.244667  66.278000   
Haiti                     30.3445  42.456143  41.151071  76.438167  40.172250   
Uganda                    23.6640  64.421429  47.922071  60.499000  60.577250   
Armenia                   93.4380  60.728143  77.907214  90.329000  57.488250   
Senegal                   30.1240  61.413750  57.519643  37.404333  44.899500   
Cyprus                    99.9310  59.192250  90.834154  97.778000  71.391333   
Mongolia                  96.8105  41.551286  64.831786  98.722333  65.205750   
Papua New Guinea          29.0085  29.557200  45.642000  59.790667  48.287000   
Botswana                  54.7840  41.411375  53.730143  89.671333  70.166750   
Brunei Darussalam         91.0000  54.708750  86.957083  99.409667  61.741333   

                              SDG6       SDG7       SDG8       SDG9  \
Country                                                               
Portugal                  84.13680  95.022667  82.155333  75.417667   
Poland                    84.91120  89.265667  84.405333  71.760000   
Belize                    75.87380  92.313500  58.470200  22.748000   
Central African Republic  45.51140  12.827500  47.698000   3.526000   
Turkey                    75.12840  89.161000  59.060667  60.301500   
Italy                     82.13500  93.847333  76.572000  83.604500   
Ireland                   87.28540  93.429333  87.328600  82.017000   
Sri Lanka                 56.62700  66.947667  76.219500  34.159167   
Vietnam                   73.13800  83.200333  72.149167  43.975667   
Mauritania                58.32540  42.221500  45.026167  17.474833   
Malta                     75.54900  95.293667  86.307750  65.619833   
Finland                   94.44060  96.486000  88.277667  94.057667   
Jordan                    57.51520  92.943000  58.027333  50.501333   
Jamaica                   73.93160  85.275667  74.816833  37.343833   
Comoros                   49.00875  42.706000  49.735000  15.946500   
El Salvador               75.80560  87.633333  63.323167  20.198833   
Chad                      39.96960   2.039000  55.779600   8.218667   
Bulgaria                  70.81480  91.049000  78.137333  52.552333   
Venezuela, RB             78.13480  92.112000  70.762800  32.317333   
Haiti                     53.76900  29.124667  55.629000  12.171500   
Uganda                    44.93380  

In [172]:
test_csv.to_csv('test.csv', index = True)

In [173]:
import json
  
# Data to be written
dictionary ={
    "nombre_alumno" : "Miguel Chamochin",
    "titulo" : "Conectando los Objetivos de Desarrollo Sostenible con el cambio climático y la transición energética",
    "tipo_ml" : "O",
    "target" : "Wealthy"
}
  
with open("info.json", "w") as outfile:
    json.dump(dictionary, outfile)

# Predicciones - SDGs_Ranks_2021

In [190]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

X = SDGs_Ranks_2021.drop('Wealthy',1)
y = SDGs_Ranks_2021['Wealthy']

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [191]:
pipeline = Pipeline(steps = [
    ("scaler", StandardScaler()), # primero escalo
    ("pca", PCA()), # segundo aplica PCA 
    ("kneighborsclassifier", KNeighborsClassifier()) # Despues un KNeighborsClassifier
])

pipeline_param = {
    'pca__n_components' :  list(range(1,17)),
    'kneighborsclassifier__n_neighbors' :  list(range(1,50))
}

gs_pipeline = GridSearchCV(pipeline,
                            pipeline_param,
                             cv = 10,
                            scoring = 'accuracy',
                            verbose = 1, # mensajes del output
                            n_jobs = -1)

grids = {"gs_pipeline": gs_pipeline}

for nombre, grid_search in grids.items():
    grid_search.fit(X_train, y_train)

Fitting 10 folds for each of 784 candidates, totalling 7840 fits


In [192]:
print(gs_pipeline.best_score_)
print(gs_pipeline.best_params_)
print(gs_pipeline.best_estimator_)

0.8890909090909093
{'kneighborsclassifier__n_neighbors': 1, 'pca__n_components': 10}
Pipeline(steps=[('scaler', StandardScaler()), ('pca', PCA(n_components=10)),
                ('kneighborsclassifier', KNeighborsClassifier(n_neighbors=1))])


In [193]:
print("Accuracy train", gs_pipeline.best_estimator_.score(X_train, y_train))
print("Accuracy test", gs_pipeline.best_estimator_.score(X_test, y_test))

Accuracy train 1.0
Accuracy test 0.6785714285714286


# CONCLUSION